File: src/213_House_Robber_II/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    //Runtime: 0 ms, faster than 100.00%
    //Memory Usage: 8.1 MB, less than 100.00%
    int rob(vector<int>& nums) {
        int size = nums.size();
        if (!size) return 0;
        
        int rob1[2], nrob1[2];
        rob1[0] = rob1[1] = nums[0];
        nrob1[0] = nrob1[1] = 0;//大意了，如果用array的形式，一定要初始化，要不然nrob1里面是垃圾数据
        
        // 完全背包问题的变种，rob1表示抢第一家，nrob1表示不抢第一家
        for (int i = 2; i < size-1; ++i) {
            // rob[i] = max(rob[i-1], rob[i-2]+nums[i])
            swap(rob1[0], rob1[1]);
            rob1[1] = max(rob1[1] + nums[i], rob1[0]);
        }
        for (int i = 1; i < size; ++i) {
            // 相同的操作给nrob1
            swap(nrob1[0], nrob1[1]);
            nrob1[1] = max(nrob1[1] + nums[i], nrob1[0]);
        }

        // 最后判断一次[size-1]这一家
        return max(rob1[1],nrob1[1]);
    }

    int rob1(vector<int>& nums) {
        int n = nums.size();
        if (n < 1) return 0;
        // 极限情况, 只有1个值; 后面的helper函数无法处理
        int rtn = nums[0];
        // 一次DP解决不了, 那就两次
        rtn = max(rtn, helper(nums, 0, n-2));//去掉最后一个房间, 用house robber I的方法
        rtn = max(rtn, helper(nums, 1, n-1));//去掉第一个房间, 用house robber I的方法
        return rtn;
    }
    
    int helper(vector<int>& nums, int bgn, int end) {
        int n = end-bgn+1;
        if (n < 1) return 0;
        vector<int> dp(n+1,0);
        dp[1] = nums[bgn];//nums[0], 得修改为当前得bgn
        for (int i = bgn+1; i <= end; ++i) {
            // for循环的i范围是正确的
            // 但是i需要转换为dp的index, 必须减去bgn
            // dp[i+1] = max(dp[i], dp[i-1] + nums[i]);
            dp[i+1-bgn] = max(dp[i-bgn], dp[i-1-bgn] + nums[i]);
        }
        return dp.back();
    }
};

int main() {
    Solution gua;
    int rtn;
    vector<int> nums2{1,2,1,1};
    rtn = gua.rob(nums2);
    cout << "expect value = 3, actual value = " << rtn << endl;    

    vector<int> nums1{1,2,3,1};
    rtn = gua.rob(nums1);
    cout << "expect value = 4, actual value = " << rtn << endl;    

    return 0;
}

File: src/213_House_Robber_II/main.py

from typing import List

class Solution:
    # Runtime: 32 ms, faster than 63.16%
    # Memory Usage: 14 MB, less than 94.12%
    def rob(self, nums: List[int]) -> int:
        # 和198题一样思路，只不过划分为多个dp数组，一定抢[0]，一定不抢[0]情况下的最大值
        size = len(nums)
        rob0 = [0] * size
        nrob0 = [0] * size
        
        # 初始化
        rob0[0] = nums[0]
        if size > 1:
            rob0[1] = nums[0]
            nrob0[1] = nums[1]
        
        # 然后分别运算
        for i in range(2,size-1): #对于rob0的更新，只到nums[size-2]
            rob0[i] = max(rob0[i-1], rob0[i-2] + nums[i])
        
        for i in range(2,size): #对于nrob0的更新，可以到nums[size-1]
            nrob0[i] = max(nrob0[i-1], nrob0[i-2] + nums[i])

        # 最后从两者里面取较大者
        return max(rob0[size-2], nrob0[size-1])

if __name__ == "__main__":
    gua = Solution()
    rtn = gua.rob(nums=[2,3,3])
    print("expect result = 3, actual result = ", rtn)

    rtn = gua.rob(nums=[1,2,3,1])
    print("expect result = 4, actual result = ", rtn)

    rtn = gua.rob(nums=[0])
    print("expect result = 0, actual result = ", rtn)


